File: frontend/src/content/core/contentPack.ts
import type { BattleConfig } from '../../components/BattleTestScreen';
import type { ItemDefinition } from '../../core/items/ItemTypes';
import type { NPCProfile } from '../../core/npc/NPCProfile';
import type { CompatibleGameState } from '../../core/state/gameState';

export interface SceneDefinition {
  id: string;
  title: string;
  metadata?: Record<string, unknown>;
}

export interface EncounterDefinition {
  id: string;
  title: string;
  battleConfig?: BattleConfig;
  metadata?: Record<string, unknown>;
}

export interface QuestDefinition {
  id: string;
  title: string;
  metadata?: Record<string, unknown>;
}

export interface ContentMigration {
  fromVersion: string;
  toVersion: string;
  migrateSave: (state: CompatibleGameState) => CompatibleGameState;
}

export interface ContentAssetManifest {
  images?: Record<string, string>;
  audio?: Record<string, string>;
}

export interface ContentPack {
  packId: string;
  version: string;
  title: string;
  description?: string;
  scenes: SceneDefinition[];
  encounters: EncounterDefinition[];
  items: ItemDefinition[];
  npcs: NPCProfile[];
  quests: QuestDefinition[];
  migrations?: ContentMigration[];
  assets?: ContentAssetManifest;
}

File: backend/content/core/content_pack.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from core.npc.npc_profile import NPCProfile


class ContentPack(BaseModel):
    packId: str
    version: str
    title: str
    description: str | None = None
    scenes: list[dict[str, Any]] = Field(default_factory=list)
    encounters: list[dict[str, Any]] = Field(default_factory=list)
    items: list[dict[str, Any]] = Field(default_factory=list)
    npcs: list[NPCProfile] = Field(default_factory=list)
    quests: list[dict[str, Any]] = Field(default_factory=list)
    migrations: list[dict[str, Any]] = Field(default_factory=list)
    assets: dict[str, Any] = Field(default_factory=dict)
